crates/flipperzero/src/furi/kernel.rs

```diff
@@ -0,0 +1,90 @@
+//! Furi Kernel primitives.
+
+use flipperzero_sys as sys;
+use ufmt::derive::uDebug;
+
+use crate::furi;
+
+/// Check if CPU is in IRQ; or kernel running and IRQ is masked.
+pub fn is_irq_or_masked() -> bool {
+    unsafe { sys::furi_kernel_is_irq_or_masked() }
+}
+
+/// Check if kernel is running.
+pub fn is_running() -> bool {
+    unsafe { sys::furi_kernel_is_running() }
+}
+
+/// Kernel lock state.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, uDebug)]
+pub enum LockState {
+    /// normal scheduling
+    Unlocked = 0,
+    /// scheduling paused
+    Locked = 1,
+}
+
+impl From<i32> for LockState {
+    fn from(value: i32) -> Self {
+        match value {
+            0 => LockState::Unlocked,
+            // Assume any non-zero value is `Locked``.
+            // This can be modified if new lock states are added in the future.
+            _ => LockState::Locked,
+        }
+    }
+}
+
+impl From<LockState> for i32 {
+    fn from(value: LockState) -> Self {
+        value as i32
+    }
+}
+
+/// Lock kernel, pause process scheduling.
+///
+/// **warning** This should never be called in an interrupt request context.
+///
+/// Returns previous lock state.
+pub fn lock() -> furi::Result<LockState> {
+    let status = sys::furi::Status::from(unsafe { sys::furi_kernel_lock() });
+
+    status.err_or_else(|status| status.0.into())
+}
+
+/// Unlock kernel, resume process scheduling.
+///
+/// **warning** This should never be called in an interrupt request context.
+///
+/// Returns previous lock state.
+pub fn unlock() -> furi::Result<LockState> {
+    let status = sys::furi::Status::from(unsafe { sys::furi_kernel_unlock() });
+
+    status.err_or_else(|status| status.0.into())
+}
+
+/// Restore kernel lock state.
+///
+/// **warning** This should never be called in an interrupt request context.
+///
+/// Returns previous lock state.
+pub fn restore_lock(state: LockState) -> furi::Result<LockState> {
+    let status = sys::furi::Status::from(unsafe { sys::furi_kernel_restore_lock(state.into()) });
+
+    status.err_or_else(|status| status.0.into())
```

Comment:
Just looking at the repetetive

```rust
let status = sys::furi::Status::from(...);
status.err_or_else(|status| status.0.into())
```

I start to feel that it's worth adding a convenience method to `Status` for this conversion such as this can be replaced with a one-liner

```rust
sys::furi::Status::into_result(...)
```

Reply:
I agree, though I need to think about how to do it in a clean way.

The challenge with `i32` status is that it can be interpreted in many ways (there's not a single mapping).

I might be overthinking it and should just provide an `into_result(self) -> Result<i32, Status>` and let the user do the mapping as desired. `ENEEDMORECAFFEINE`. 🍵

Reply:
Since this might involving some more refactoring will address in a later PR.